defi/single-token-amm-dex/src/lib.rs

```diff
@@ -0,0 +1,914 @@
+                    new_price = total_candy*price/(total_candy+candy_amnt);
+                }
+
+                new_price
+            }
+
+            // Calculate candy output amount.
+            fn candy_sum(
+                &mut self, 
+                amnt_pri: Decimal, 
+                addr_pri: Address, 
+                addr_sec:Address, 
+                flag: Decimal
+            )-> Decimal {
+                let mut amount: Decimal = amnt_pri;
+
+                let price_in: Decimal = CandyDex::candyprice(self, addr_pri);
+                let price_out : Decimal = CandyDex::candyprice(self, addr_sec);
+
+                let tot_amnt_out = self.candy_vaults.get(&addr_sec).unwrap().amount();
+                let candy_out_amnt = amnt_pri*price_in/price_out;
+                
+                if flag == dec!(2) {
+                    amount = amnt_pri/price_in;
+                }
+
+                if flag == dec!(1) {
+                    let price_new = price_out*tot_amnt_out/(tot_amnt_out+candy_out_amnt);
+                    amount = price_new*candy_out_amnt/price_out;
+                } else {
+                    let price_new = 
+                        CandyDex::price_mod(self, amount, addr_pri, price_in, dec!(1));
+                    let (name,symbol,_price) = self.candy_map.get(&addr_pri).unwrap();
+                    let name_str = name.to_string();
+                    let symbol_str = symbol.to_string();
+                    self.candy_map.insert(addr_pri,(name_str,symbol_str,price_new));
+                    amount = amnt_pri/price_new;
+                }
+
+                amount
+            }
+
+            // Adjust buying exact candy amount neutralizing protocol fee incidence on final amount.
+            fn adjust_fee(&mut self, amount_in: Decimal ) -> Decimal {
+                let hundred = dec!(100);
+                let amount_out = amount_in*hundred/(hundred-self.fee);
+                
+                amount_out
+            }
+
+            // Set protocol fee function whom only protocol owner can succesfully call.
+        #[auth(owner_badge)]
+        pub fn set_fee(&mut self, prtcl_fee: Decimal) {
+            assert!(prtcl_fee >= dec!(0) && prtcl_fee <= dec!(1)," Let's pass a fee in thousandths! ");
+            
+            self.fee = prtcl_fee;
+            info!(" Protocol fee set to {}% ", self.fee);
+        }
+
+            // Claim accrued XRD fee function whom only protocol owner can succesfully call.
+        #[auth(owner_badge)]
+        pub fn claim_xrd_fee(&mut self) -> Bucket {
+            info!(" Fee value {} XRD ", self.xrd_fee);
+            
+            let xrd_amount = self.collected_xrd.amount();
+
+            let xrd_output: Bucket = self.collected_xrd.take(self.xrd_fee);
+            
+            self.xrd_claimed += xrd_amount-self.collected_xrd.amount();
+            info!(" Protocol fee claimed {} XRD ", self.xrd_claimed);
+            
+            xrd_output
```

Comment:
This can be simplified like this:

```suggestion
            info!(" Fee value {} XRD ", self.xrd_fee);
            
            self.xrd_claimed += self.collected_xrd.amount();
            info!(" Protocol fee claimed {} XRD ", self.xrd_claimed);
            self.collected_xrd.take_all()
```

Reply:
Undone with objection as protocol is specifically designed to consent Owner to withdraw only accrued fee amount from XRD Vault.
Remaing XRD liquidity is utilized by methods invoked by users who desire sell their Candies in exchange of XRD 
as well as reimburse liquidity providers who desire to withdraw an amount of provided Candy higher than the disposable one in protocol reserve, in case overnominated Candy had been overbuyed by other users.

Anyway thanks a lot as to get me back to this piece of code as I found a bug: I forgot to reset " self.xrd_fee " each time fee is withdraw, I've also simplified a bit:

            info!(" Fee value {} XRD ", self.xrd_fee);
        
            let xrd_output: Bucket = self.collected_xrd.take(self.xrd_fee);

            self.xrd_fee = dec!(0);

            self.xrd_claimed += xrd_output.amount();
            info!(" Protocol fee claimed {} XRD ", self.xrd_claimed);
            
            xrd_output